crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -194,6 +194,90 @@ where
             Ok(None)
         }
     }
+
+    /// Returns a list of entries in the given transaction range.
+    ///
+    /// First, the entries are fetched from the database using the provided closure. When there's no
+    /// database entries left, but the range is still not exhausted, the entries are fetched
+    /// from the in-memory state using the provided closure.
+    ///
+    /// - `query_database` is called with the transaction range
+    /// - `query_in_memory` is called with the block state and the index of the transaction in the
+    ///   block
+    ///
+    /// The entries are returned in the order of the transaction numbers.
+    fn entries_by_tx_range<T>(
+        &self,
+        range: impl RangeBounds<TxNumber>,
+        query_database: impl Fn(RangeInclusive<TxNumber>) -> ProviderResult<Vec<T>>,
+        query_in_memory: impl Fn(&BlockState, usize) -> ProviderResult<T>,
+    ) -> ProviderResult<Vec<T>> {
+        let provider = self.database.provider()?;
+
+        let (start_tx_number, end_tx_number) = self.convert_range_bounds(range, || u64::MAX);
+        let range = start_tx_number..=end_tx_number;
+
+        let last_database_block_number = provider.last_block_number()?;
+
+        if start_tx_number <= last_database_block_number &&
+            (matches!(range.end_bound(), Bound::Included(&n) if n < last_database_block_number) ||
+                matches!(range.end_bound(), Bound::Excluded(&n) if n <= last_database_block_number))
```

Comment:
Could you pls add a comment here? not sure why `<` is used in the guard when end bound is `Included` and `<=` when it is included

Reply:
yep, fixed and added a comment